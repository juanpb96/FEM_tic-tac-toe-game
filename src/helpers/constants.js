export const ASSETS_PATH = `${process.env.PUBLIC_URL}/assets`;