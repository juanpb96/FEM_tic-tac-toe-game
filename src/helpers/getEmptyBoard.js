const EMPTY_BOARD = [
    [null, null, null],
    [null, null, null],
    [null, null, null],
];

export const getEmptyBoard = () => EMPTY_BOARD.map(row => row.slice(0));