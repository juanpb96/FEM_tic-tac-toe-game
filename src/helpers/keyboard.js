export const isKeyEnterPressed = key => key && ['Enter', ' '].includes(key);