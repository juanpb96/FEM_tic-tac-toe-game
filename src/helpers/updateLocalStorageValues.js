import { STORAGE,USER } from '../types/types';

const {
    lsCpuScore,
    lsPlayerScore,
    lsP1Score,
    lsP2Score,
    lsTiedScore,
} = STORAGE;

const updateStorageScore = (key) => {
    const currentScore = +localStorage.getItem(key);
        
    localStorage.setItem(
        key,
        currentScore ? currentScore + 1 : 1
    );
};

export const updateScore = (winner = '') => {
    switch (winner) {
        case USER.cpu:
            updateStorageScore(lsCpuScore);
            break;
        case USER.player:
            updateStorageScore(lsPlayerScore);
            break;
        case USER.p1:
            updateStorageScore(lsP1Score);
            break;
        case USER.p2:
            updateStorageScore(lsP2Score);
            break;
        default:
            updateStorageScore(lsTiedScore);
            break;
    }
};