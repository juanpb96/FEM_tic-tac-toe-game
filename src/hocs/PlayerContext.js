import { createContext } from 'react';

export const PlayerCoxtext = createContext(null);