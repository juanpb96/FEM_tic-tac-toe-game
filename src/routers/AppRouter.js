import { Routes, Route } from 'react-router-dom';
import { Game } from '../components/Game/Game';
import { NewGame } from '../components/NewGame/NewGame';

export const AppRouter = () => {
    return (
        <Routes>
            <Route path='/' element={<Game />} />
            <Route path='/new-game' element={<NewGame />} />
            <Route path='*' element={<Game />} />
        </Routes>
    );
};
