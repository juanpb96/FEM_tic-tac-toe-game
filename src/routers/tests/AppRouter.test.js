import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AppRouter } from '../AppRouter';
import { GameContext } from '../../hocs/GameContext';
import { getEmptyBoard } from '../../helpers/getEmptyBoard';

describe('Test <AppRouter />', () => {
    const gameState = {
        currentPlayer: 'X',
        board: getEmptyBoard(),
        turnCounter: 1,
        isGameOver: false,
        isCpuFirstMove: true
    };

    const dispatch = jest.fn();

    test('should render Game route correctly', () => { 
        const { container } = render(
            <GameContext.Provider value={{ gameState, dispatch }}>
                <MemoryRouter initialEntries={['/']}>
                    <AppRouter />
                </MemoryRouter>
            </GameContext.Provider>
        );

        expect(container).toMatchSnapshot();
        expect(screen.getByLabelText('restart')).toBeInTheDocument();
    });

    test('should render NewGame route correctly', () => { 
        const { container } = render(
            <GameContext.Provider value={{ gameState, dispatch }}>
                <MemoryRouter initialEntries={['/new-game']}>
                    <AppRouter />
                </MemoryRouter>
            </GameContext.Provider>
        );

        expect(container).toMatchSnapshot();
        expect(screen.getByText('REMEMBER : X GOES FIRST')).toBeInTheDocument();
    });   
});