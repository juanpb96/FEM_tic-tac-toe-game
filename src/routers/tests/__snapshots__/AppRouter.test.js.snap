// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <AppRouter /> should render Game route correctly 1`] = `
<div>
  <div
    class="[ page-content align-self-start mt-6 tablet:align-self-center tablet:mt-0 ]"
  >
    <header
      class="grid grid-cols-3 align-center gap-5 mb-16 tablet:mb-5"
    >
      <h1
        class="sr-only"
      >
        Tic tac toe
      </h1>
      <img
        alt=""
        src="/assets/logo.svg"
      />
      <section
        class="[ turn ][ flex gap-2.5 box-navy br-1.5 tablet:br-2.5 ]"
        data-type="secondary"
      >
        <svg
          aria-label="X"
          fill="none"
          height="20"
          role="img"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M19.7231 3.30608L16.6939 0.276913C16.3247 -0.0923043 15.7261 -0.0923043 15.3569 0.276913L10 5.63378L4.64314 0.276913C4.27392 -0.0923043 3.6753 -0.0923043 3.30608 0.276913L0.276913 3.30608C-0.0923043 3.6753 -0.0923043 4.27392 0.276913 4.64314L5.63378 10L0.276913 15.3569C-0.0923043 15.7261 -0.0923043 16.3247 0.276913 16.6939L3.30608 19.7231C3.6753 20.0923 4.27392 20.0923 4.64314 19.7231L10 14.3662L15.3569 19.7231C15.7261 20.0923 16.3247 20.0923 16.6939 19.7231L19.7231 16.6939C20.0923 16.3247 20.0923 15.7261 19.7231 15.3569L14.3662 10L19.7231 4.64314C20.0923 4.27392 20.0923 3.6753 19.7231 3.30608Z"
            fill="#A8BFC9"
            fill-rule="evenodd"
          />
        </svg>
        <h2
          class="fs-3.5 color-silver tablet:fs-4"
        >
          TURN
        </h2>
      </section>
      <button
        aria-label="restart"
        class="[ button-restart ][ justify-self-end br-1.5 tablet:br-2.5 ]"
        data-type="secondary"
      >
        <svg
          fill="currentColor"
          height="20"
          role="presentation"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.524 0h-1.88a.476.476 0 0 0-.476.499l.159 3.284A9.81 9.81 0 0 0 9.835.317C4.415.317-.004 4.743 0 10.167.004 15.597 4.406 20 9.835 20a9.796 9.796 0 0 0 6.59-2.536.476.476 0 0 0 .019-.692l-1.348-1.349a.476.476 0 0 0-.65-.022 6.976 6.976 0 0 1-9.85-.63 6.987 6.987 0 0 1 .63-9.857 6.976 6.976 0 0 1 10.403 1.348l-4.027-.193a.476.476 0 0 0-.498.476v1.881c0 .263.213.476.476.476h7.944A.476.476 0 0 0 20 8.426V.476A.476.476 0 0 0 19.524 0Z"
            fill=""
          />
        </svg>
      </button>
    </header>
    <main
      class="[ board ][ mb-5 ]"
    >
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 tablet:br-4 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
    </main>
    <footer
      class="[ flex space-between align-center gap-5 ]"
    >
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-light-blue p-3 br-2.5 tablet:br-4 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          X ()
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        >
          0
        </p>
      </section>
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-silver p-3 br-2.5 tablet:br-4 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          TIES
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        />
      </section>
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-light-yellow p-3 br-2.5 tablet:br-4 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          O ()
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        >
          0
        </p>
      </section>
    </footer>
  </div>
</div>
`;

exports[`Test <AppRouter /> should render NewGame route correctly 1`] = `
<div>
  <main
    class="[ flow ][ page-content ]"
  >
    <h1
      class="sr-only"
    >
      Tic tac toe
    </h1>
    <img
      alt=""
      class="logo"
      src="/assets/logo.svg"
    />
    <section
      aria-labelledby="mark-label"
      class="selection-container"
      role="radiogroup"
    >
      <h2
        class="[ fs-4 color-silver letter-xs mb-6 ]"
        id="mark-label"
      >
        PICK PLAYER 1â€™S MARK
      </h2>
      <div
        class="[ flex bg-dark-navy mb-4 px-2 py-2.5 br-2.5 ]"
      >
        <div
          aria-checked="false"
          aria-labelledby="label-x"
          class="[ pick-mark ][ flex flex-1 flex-center py-3 br-2.5 ]"
          role="radio"
          tabindex="0"
        >
          <span
            class="sr-only"
            id="label-x"
          >
            X
          </span>
          <svg
            fill="currentColor"
            height="32"
            role="presentation"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M31.5569 5.28973L26.7103 0.443061C26.1195 -0.147687 25.1617 -0.147687 24.571 0.443061L16 9.01404L7.42902 0.443061C6.83827 -0.147687 5.88048 -0.147687 5.28973 0.443061L0.443061 5.28973C-0.147687 5.88048 -0.147687 6.83827 0.443061 7.42902L9.01404 16L0.443061 24.571C-0.147687 25.1617 -0.147687 26.1195 0.443061 26.7103L5.28973 31.5569C5.88048 32.1477 6.83827 32.1477 7.42902 31.5569L16 22.986L24.571 31.5569C25.1617 32.1477 26.1195 32.1477 26.7103 31.5569L31.5569 26.7103C32.1477 26.1195 32.1477 25.1617 31.5569 24.571L22.986 16L31.5569 7.42902C32.1477 6.83827 32.1477 5.88048 31.5569 5.28973Z"
              fill=""
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          aria-checked="true"
          aria-labelledby="label-o"
          class="[ pick-mark ][ flex flex-1 flex-center py-3 br-2.5 ]"
          role="radio"
          tabindex="0"
        >
          <span
            class="sr-only"
            id="label-o"
          >
            O
          </span>
          <svg
            fill="currentColor"
            height="32"
            role="presentation"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M31.7412 15.8706C31.7412 7.10551 24.6357 0 15.8706 0C7.10551 0 0 7.10551 0 15.8706C0 24.6357 7.10551 31.7412 15.8706 31.7412C24.6357 31.7412 31.7412 24.6357 31.7412 15.8706ZM9.4048 15.8706C9.4048 12.2996 12.2996 9.4048 15.8706 9.4048C19.4416 9.4048 22.3364 12.2996 22.3364 15.8706C22.3364 19.4416 19.4416 22.3364 15.8706 22.3364C12.2996 22.3364 9.4048 19.4416 9.4048 15.8706Z"
              fill=""
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <p
        class="[ fs-3.5 color-silver opacity ]"
      >
        REMEMBER : X GOES FIRST
      </p>
    </section>
    <div
      class="[ flex flex-col gap-4 tablet:gap-5 ]"
    >
      <button
        class="[ button ][ yellow ]"
        data-type="primary"
      >
        New Game (VS CPU)
      </button>
      <button
        class="[ button ][ blue ]"
        data-type="primary"
      >
        New Game (VS Player)
      </button>
    </div>
  </main>
</div>
`;
